group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -3500,6 +3503,59 @@ public void maybeDeleteGroup(String groupId, List<Record> records) {
         }
     }
 
+    /**
+     * A group can be upgraded offline if it's a classic group and empty.
+     *
+     * @param groupId The group to be validated.
+     * @return true if the offline upgrade is valid.
+     */
+    private boolean validateOfflineUpgrade(String groupId) {
+        Group group = groups.get(groupId);
+
+        if (group == null || group.type() == CONSUMER) {
+            return false;
+        }
+
+        ClassicGroup classicGroup = (ClassicGroup) group;
+        if (!classicGroup.isEmpty()) {
+            return false;
+        } else {
+            return true;
+        }
+    }
+
+    /**
+     * Upgrade the empty classic group to a consumer group if it's valid.
+     *
+     * @param groupId       The group id to be updated.
+     * @param records       The list of records to delete the classic group and create the consumer group.
+     * @param isSimpleGroup The boolean indicating whether the group to be updated is a simple group.
+     */
+    public void maybeUpgradeEmptyGroup(String groupId, List<Record> records, boolean isSimpleGroup) {
+        if (validateOfflineUpgrade(groupId)) {
+            final long currentTimeMs = time.milliseconds();
+            ClassicGroup classicGroup = getOrMaybeCreateClassicGroup(groupId, false);
+            int groupEpoch = classicGroup.generationId();
+
+            // Replace the classic group with a new consumer group.
+            ConsumerGroup consumerGroup = getOrMaybeCreateConsumerGroup(groupId, true);
+            // We don't create the tombstone because the replay will remove the newly created consumer group.
```

Comment:
I think that we must write a tombstone for the old group. As you said, the map will be updated based on the records with the new group. However, we also need to compact the old record for the group and the only way to do it is to write a tombstone.